src/extensions/default/JavaScriptCodeHints/node/TernNodeDomain.js

```diff
@@ -0,0 +1,826 @@
+        name: path,
+        offsetLines: 0,
+        text: ""};
+}
+
+/**
+ * Build an object that can be used as a request to tern.
+ *
+ * @param {{type: string, name: string, offsetLines: number, text: string}} fileInfo
+ * - type of update, name of file, and the text of the update.
+ * For "full" updates, the whole text of the file is present. For "part" updates,
+ * the changed portion of the text. For "empty" updates, the file has not been modified
+ * and the text is empty.
+ * @param {string} query - the type of request being made
+ * @param {{line: number, ch: number}} offset -
+ */
+function buildRequest(fileInfo, query, offset) {
+    query = {type: query};
+    query.start = offset;
+    query.end = offset;
+    query.file = (fileInfo.type === MessageIds.TERN_FILE_INFO_TYPE_PART) ? "#0" : fileInfo.name;
+    query.filter = false;
+    query.sort = false;
+    query.depths = true;
+    query.guess = true;
+    query.origins = true;
+    query.types = true;
+    query.expandWordForward = false;
+    query.lineCharPositions = true;
+    query.docs = true;
+    query.urls = true;
+
+    var request = {query: query, files: [], offset: offset, timeout: inferenceTimeout};
+    if (fileInfo.type !== MessageIds.TERN_FILE_INFO_TYPE_EMPTY) {
+        request.files.push(fileInfo);
+    }
+
+    return request;
+}
+
+/**
+ * Get definition location
+ * @param {{type: string, name: string, offsetLines: number, text: string}} fileInfo
+ * - type of update, name of file, and the text of the update.
+ * For "full" updates, the whole text of the file is present. For "part" updates,
+ * the changed portion of the text. For "empty" updates, the file has not been modified
+ * and the text is empty.
+ * @param {{line: number, ch: number}} offset - the offset into the
+ * file for cursor
+ */
+function getJumptoDef(fileInfo, offset) {
+    var request = buildRequest(fileInfo, "definition", offset);
+    // request.query.typeOnly = true;       // FIXME: tern doesn't work exactly right yet.
+
+    try {
+        ternServer.request(request, function (error, data) {
+            if (error) {
+                _log("Error returned from Tern 'definition' request: " + error);
+                self.postMessage({type: MessageIds.TERN_JUMPTODEF_MSG, file: fileInfo.name, offset: offset});
+                return;
+            }
+            var response = {type: MessageIds.TERN_JUMPTODEF_MSG,
+                                  file: fileInfo.name,
+                                  resultFile: data.file,
+                                  offset: offset,
+                                  start: data.start,
+                                  end: data.end
+                                 };
+
+            request = buildRequest(fileInfo, "type", offset);
+            // See if we can tell if the reference is to a Function type
+            ternServer.request(request, function (error, data) {
+                if (!error) {
```

Comment:
This is just a check to tell the type. If we can't determine the type using tern we send the original response back.